[REVIEW] Fix gather null mask allocation for invalid index

kaatish opened the pull request:
Gather has been updated to produce invalid rows in a table if an invalid index is contained in the gather map and if `ignore_out_of_bounds` is true.

<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

jrhemstad commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -338,11 +377,15 @@ gather(table_view const& source_table, MapIterator gather_map_begin,
 
   rmm::device_vector<bitmask_type*> masks(host_masks);
 
+  cudaDeviceSynchronize();
```
Definitely no device syncs allowed.

kaatish replied:
Debugging artifact

jrhemstad commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -338,11 +377,15 @@ gather(table_view const& source_table, MapIterator gather_map_begin,
                                                           masks.data().get(),
                                                           destination_table->num_rows(),
                                                           valid_counts.data().get());
+  cudaDeviceSynchronize();
```
No device syncs.

jrhemstad commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -116,24 +116,38 @@ struct column_gatherer_impl
     std::unique_ptr<column> operator()(column_view const& source_column,
                                        MapIterator gather_map_begin,
                                        MapIterator gather_map_end,
+                                       cudf::experimental::mask_allocation_policy policy,
```
This doesn't make sense to me. Shouldn't the state of the output null mask depend only on the existence of nulls in the source column?

jrhemstad replied:
Okay, I understand. If `ignore_out_of_bounds` is true, you need to have a null mask allocated regardless of the presence of nulls in the source. The `mask_allocation_policy` parameter feels very awkward in this API. Instead, why not just always allocate a bitmask for the output column if `ignore_out_of_bounds == true`?

kaatish replied:
That would definitely make it easier. However, my motivation was to avoid allocation of masks if all bits are supposed to be valid.

jrhemstad replied:
I'd prefer having an unnecessary allocation to adding a `policy` argument to the `gather` API. Otherwise it introduces possibility for user error like setting `policy == NEVER` when `ignore_out_of_bounds == true` and there _are_ out of bound indices.

kaatish replied:
Ok that's a lot easier. Will fix it

harrism commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -321,7 +334,7 @@ gather(table_view const& source_table, MapIterator gather_map_begin,
 
   auto bitmask_kernel =
-    ignore_out_of_bounds ? gather_bitmask_kernel<true, decltype(gather_map_begin)> : gather_bitmask_kernel<false, decltype(gather_map_begin)>;
+    nullify_out_of_bounds ? gather_bitmask_kernel<true, decltype(gather_map_begin)> : gather_bitmask_kernel<false, decltype(gather_map_begin)>;
```
A bit long line.

```suggestion
    nullify_out_of_bounds ? gather_bitmask_kernel<true, decltype(gather_map_begin)> : 
                            gather_bitmask_kernel<false, decltype(gather_map_begin)>;
```

trevorsm7 commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -116,27 +116,40 @@ struct column_gatherer_impl
+              stream));
+      }
+
+      if (nullify_out_of_bounds) {
```
Could this be combined with the if above?

trevorsm7 commented on cpp/tests/copying/gather_tests.cu:
```diff
@@ -55,6 +56,31 @@ TYPED_TEST(GatherTest, GatherDetailDeviceVectorTest) {
+TYPED_TEST(GatherTest, GatherDetailInvalidIndexTest) {
+  constexpr cudf::size_type source_size{1000};
+
+  auto data = cudf::test::make_counting_transform_iterator(0, [](auto i){return i;});
```
Could this be a thrust::counting_iterator?

jrhemstad commented:
rerun tests

harrism commented:
Some of the out of bounds strings gather tests are failing.

kaatish commented:
> Some of the out of bounds strings gather tests are failing.

I have fixed that. The issue was that strings gather did not calculate the null masks. I have replaced that with the generic gather and those tests pass.

The current problem I am facing is in scatter to tables tests. I believe scatter uses gather internally at some point. Now that we are always allocating null masks when 'nullify_out_of_bounds` is true, those tests are failing because the expected tables and the result tables do not have the same `nullable()` flag.

I am trying to fix the tests but I don't have a straightforward solution yet.
